notify/templates.go

```diff
@@ -0,0 +1,240 @@
+	paths := make([]string, 0)
+	for _, name := range am.templates {
+		if name == c.Name {
+			// Skip the existing template of the same name as we're going to parse the one for testing instead.
+			continue
+		}
+		paths = append(paths, filepath.Join(am.workingDirectory, name))
+	}
+
+	// Parse current templates.
+	var newTextTmpl *tmpltext.Template
+	var captureTemplate template.Option = func(text *tmpltext.Template, _ *tmplhtml.Template) {
+		newTextTmpl = text
+	}
+	newTmpl, err := templates.FromGlobs(paths, captureTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse test template.
+	_, err = newTextTmpl.Parse(c.Template)
+	if err != nil {
+		// This shouldn't happen since we already parsed the template above.
+		return nil, err
+	}
+
+	externalURL, err := url.Parse(am.ExternalURL())
+	if err != nil {
+		return nil, err
+	}
+	newTmpl.ExternalURL = externalURL
+
+	// Prepare the context.
+	alerts := v2.OpenAPIAlertsToAlerts(c.Alerts)
+	ctx = notify.WithReceiverName(ctx, "test receiver")
+	ctx = notify.WithGroupLabels(ctx, model.LabelSet{"group_label": "group_label_value"})
+
+	var tmplErr error
+	templater, _ := templates.TmplText(ctx, newTmpl, alerts, am.logger, &tmplErr)
+
+	// Iterate over each definition in the template and evaluate it.
+	var results TestTemplatesResults
+	for _, def := range definitions {
+		s := fmt.Sprintf(`{{ template "%s" . }}`, def)
+		val := templater(s)
+		if tmplErr != nil {
+			results.Errors = append(results.Errors, TestTemplatesErrorResult{
+				Name:    def,
+				Kind:    ExecutionError,
+				Message: tmplErr.Error(),
+			})
+			tmplErr = nil
+			continue
+		}
+
+		results.Results = append(results.Results, TestTemplatesResult{
+			Name: def,
+			Text: val,
+		})
+	}
+
+	return &results, nil
+}
+
+// parseTestTemplate parses the test template and returns the top-level definitions that should be interpolated as results.
+func parseTestTemplate(name string, template string) ([]string, error) {
+	tmpl, err := tmpltext.New(name).Parse(template)
+	if err != nil {
+		return nil, err
+	}
+
+	topLevel, err := findTopLevelTemplates(tmpl)
+	if err != nil {
+		return nil, err
+	}
+
+	return topLevel, nil
+}
+
+// FindTopLevelTemplates returns the top-level definitions of the given template.
+func findTopLevelTemplates(tmpl *tmpltext.Template) ([]string, error) {
+	// We need to find the names of all defined templates and subtract
+	// the names of all executed templates to find the set of templates that
+	// should be tested.
+	definedTmpls := make([]*tmpltext.Template, 0)
+	for _, def := range tmpl.Templates() {
+		// Check defined templates for an empty outer wrapper template.
+		// This can happen if the template filename does not match the name of any template definition.
+		// Remove if it exists.
+		if def.Name() == tmpl.ParseName && def.Root.Nodes == nil && def.Root.Pos == 0 {
```

Comment:
Can you explain the significance of `Nodes == nil` and `Pos == 0` as I'm not sure I understand why it's needed, but happy to have it explained to me.

I know that `len(Nodes) == 0` when the template contains just `{{ define "name" }}{{ end }}` blocks, but what about `Pos`?

Reply:
If we left out `pos==0`, it wouldn't make that much difference but technically if they had an empty template defined with the same name as the file it would be left out accidentally, ex:

filename=`empty`
```
{{ define "slack.title" }}{{ template "default.title" . }}{{ end }}{{ define "empty" }}{{ end }}
```

Would return just `slack.title` in the response.

Reply:
I kind of understand, but I'm also kind of confused. 

If we take the example from [here](https://github.com/grobinson-grafana/template-test-example-1/blob/main/template_test.go#L19-L20) the template name is foo and it contains nodes for the `{{ range .Alerts }}{{ end }}`:

```go
tmplName: "foo",
tmplText: `{{ range .Alerts }}{{ end }}`,
```

If we do change tmplText to define foo again, the define in the tmplText just replaces the foo in `template.New()` such that the parsed template is the exact same as the example above:

```go
tmplName: "foo",
tmplText: `{{ define "foo" }}{{ range .Alerts }}{{ end }}{{ end }}`,
```

If we change tmplText to the following we have a template called *foo* that has no nodes because it just contains nested templates.

```go
tmplName: "foo",
tmplText: `{{ define "bar" }}{{ range .Alerts }}{{ end }}{{ end }}`,
```

And last if we do this we have a template called *foo* that contains a `NodeTemplate` node:

```go
tmplName: "foo",
tmplText: `{{ define "bar" }}{{ range .Alerts }}{{ end }}{{ end }}{{ template "bar" . }}`,
```

So I'm still a little confused about the need for `def.Name() == tmpl.ParseName && def.Root.Nodes == nil`. Which of the examples supposed is it supposed to work with?